File: 02-passwordPhilosophy/partOne.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

type Password struct {
	plaintext    string
	targetLetter rune
	minRepeats   int
	maxRepeats   int
}

var parserRegex = regexp.MustCompile(`(?m)(\d+)-(\d+) ([a-z]): (.+)`)

func main() {
	inputBytes, _ := ioutil.ReadFile("input.txt")
	inputSlice := strings.Split(strings.TrimSpace(string(inputBytes)), "\n")

	var passwords []Password
	for _, line := range inputSlice {
		matches := parserRegex.FindAllStringSubmatch(line, -1)
		miR, _ := strconv.Atoi(matches[0][1])
		maR, _ := strconv.Atoi(matches[0][2])
		passwords = append(passwords, Password{
			plaintext:    matches[0][4],
			targetLetter: rune(matches[0][3][0]),
			minRepeats:   miR,
			maxRepeats:   maR,
		})
	}

	var num_valid_passwords int

	for _, password := range passwords {
		var target_letter_count int
		for _, char := range password.plaintext {
			if char == password.targetLetter {
				target_letter_count += 1
			}
		}

		if (target_letter_count >= password.minRepeats) && (target_letter_count <= password.maxRepeats) {
			num_valid_passwords += 1
		}
	}

	fmt.Printf("There are %d valid passwords.\n", num_valid_passwords)
}

File: 02-passwordPhilosophy/python/partOne.py
import re

from common import *

def partOne(instr:str) -> int:
    input_string = parse(instr)

    class Password:
        plaintext: str
        target_letter: str
        min_repeats: int
        max_repeats: int

        def __init__(self, plaintext:str, target_letter:str, min_repeats:int, max_repeats:int) -> None:
            self.plaintext = plaintext
            self.target_letter = target_letter
            self.min_repeats = min_repeats
            self.max_repeats = max_repeats

    parser_regex = r"(\d+)-(\d+) ([a-z]): (.+)"

    passwords = []

    for line in input_string:
        m = re.match(parser_regex, line)
        passwords.append(Password(m.group(4), m.group(3), int(m.group(1)), int(m.group(2))))

    num_valid_passwords = 0

    for password in passwords:
        target_letter_count = 0
        for char in password.plaintext:
            if char == password.target_letter:
                target_letter_count += 1

        if password.min_repeats <= target_letter_count <= password.max_repeats:
            num_valid_passwords += 1
    
    return num_valid_passwords
